src/QMCDrivers/DriftOperators.h

```diff
@@ -157,22 +139,45 @@ inline T setScaledDriftPbyPandNodeCorr(T tau,
  * @param qf quantum forces
  * @param drift scaled quantum forces
  * @param return correction term
+ *
+ * Fill the drift vector one particle at a time (pbyp).
+ *
+ * The naive drift is tau/mass*grad_psi_over_psi,
+ *  namely the log derivative of the guiding wavefunction; tau is timestep; mass is particle mass
+ * The naive drift diverges at a node, causing persistent configurations.
+ * The norm of the drift vector should be limited in two ways:
+ *  1. Umrigar: suppress drift divergence to mimic wf divergence
+ *  2. Ceperley: limit max drift rate to diffusion rate -> set Umrigar "a" parameter
+ * The choice of drift vector does not affect VMC correctness
+ *  so long as the proposal probabilities are correctly calculated.
+ * The choice of drift vector changes the DMC Green's function. BE CAREFUL!
+ *
+ * T should be either float or double
+ * T1 may be real or complex
+ * D should be the number of spatial dimensions (int)
  */
 template<class T, class T1, unsigned D>
 inline T setScaledDriftPbyPandNodeCorr(T tau_au, const std::vector<T>& massinv,
                                        const ParticleAttrib<TinyVector<T1,D> >& qf,
                                        ParticleAttrib<TinyVector<T,D> >& drift)
 {
-  T norm=0.0, norm_scaled=0.0, vsq;
+  T vsq, tau_over_mass, sc;    // temp. variables to be assigned
+  T norm=0.0, norm_scaled=0.0; // variables to be accumulated
   for(int iat=0; iat<massinv.size(); ++iat)
   {
-    T tau=tau_au*massinv[iat];
-    convert(dot(qf[iat],qf[iat]),vsq);
-    //T vsq=dot(qf[iat],qf[iat]);
-    T sc=(vsq<std::numeric_limits<T>::epsilon())? tau:((-1.0+std::sqrt(1.0+2.0*tau*vsq))/vsq);
+    // effectively set the Umrigar "a" parameter to particle mass
+    T tau_over_mass = tau_au*massinv[iat]; // !!!! assume timestep is scaled by mass
+    // save real part of wf log derivative in drift
+    convert(qf[iat],drift[iat]);
+    vsq = dot(drift[iat],drift[iat]);
+    // calculate drift scalar "sc" of Umrigar, JCP 99, 2865 (1993); eq. (34) * tau
+    // use naive drift if vsq may cause numerical instability in the denominator
+    vsq = (vsq < std::numeric_limits<T>::epsilon()) ? tau_over_mass : vsq;
+    sc  = (-1.0+std::sqrt(1.0+2.0*tau_over_mass*vsq))/vsq;
```

Comment:
@Paul-St-Young 
line 175, 176: I don't see a reason to rewrite them since the result is basically the same.

Reply:
@ye-luo do you mean the comment "The naive drift diverges at a node..."?

Reply:
I mean the source line not the comment.

Reply:
@ye-luo indeed the functionality is unchanged. I added the `sc` variable so that the meaning of `vsq` is definitively "the gradient squared". I believe the extra `sc` variable improves the readability of the code. The separation of `sc` and `vsq` also elucidates the purpose of the ternary.

Reply:
@ye-luo I'm sorry, I was confused. Yes the two lines you mentioned should be merged. It has been a long day ...

I will change the code according to your suggestions ASAP

Reply:
The problem is that you only changed one routine but all the others are still written in the old way. I prefer them consistent.

Reply:
@ye-luo yes, I agree consistency is important. I have updated all the routines in my last commit. Please take a look.